src/box/lua/config/init.lua

```diff
@@ -57,7 +58,16 @@ end
 
 local methods = {}
 local mt = {
-    __index = methods,
+    __index = function(_, index)
```

Comment:
Can we avoid a function here? It breaks tab-completion.

If we can't, can we also define a tab-completion hook (`__autocomplete`)?

Reply:
Well, I've changed the code a bit:

```
--- a/src/box/lua/config/init.lua
+++ b/src/box/lua/config/init.lua
@@ -1,4 +1,5 @@
 local fio = require('fio')
+local fun = require('fun')
 local instance_config = require('internal.config.instance_config')
 local cluster_config = require('internal.config.cluster_config')
 local configdata = require('internal.config.configdata')
@@ -37,6 +38,7 @@ local function load_extras()
 end
 
 local extras = load_extras()
+local extra_methods = extras ~= nil and extras.export or {}
 
 -- {{{ Helpers
 
@@ -58,16 +60,11 @@ end
 
 local methods = {}
 local mt = {
-    __index = function(_, index)
-        if extras ~= nil and extras.export ~= nil and
-           extras.export[index] ~= nil then
-            return extras.export[index]
-        end
-
-        if methods[index] ~= nil then
-            return methods[index]
-        end
+    __index = setmetatable(methods, {__index = extra_methods}),
+    __autocomplete = function(self)
+        return fun.tomap(fun.chain(fun.iter(methods), fun.iter(extra_methods)))
     end,
+
     __serialize = filter_out_private_fields,
 }
```

I've tried to "merge" extra_methods and methods to use them as a table though it was problematic due to complicated loading process thus I've overridden the `__autocomplete`

What do you think now?

Reply:
Can't we just do something like this in the EE code?

```lua
local function initialize(config_module)
    <...>
    local mt = getmetatable(config_module)
    local methods = mt.__index
    methods.foo = foo
end
```

Or add `:register_method()` into CE code and add the following into EE?

```lua
local function initialize(config_module)
    <...>
    config_module:register_method('foo', foo)
end
```

It seems that we can write something simple here, because the problem doesn't look complicated.

Reply:
I don't like that approach since it requires config to be provided to access any methods added in EE (since `:initialize()` should be called)

Technically, I can add extra hook `pre_init()` that is called independently without config initialization if you insist. Though, I don't see any explicit pros comparing to the existing in it.

Reply:
I see now, the `extras.initialize(config_module)` is called only in the YAML config startup flow. OK.

I still think that your way to add a couple of methods into a metatable involves too much programming :)

Can't we have a module in EE that does something like the following?

```lua
-- my_ee_module.lua
local config = require('config')

local mt = getmetatable(config)
local methods = mt.__index

function methods.foo(self, <...>)
    <...>
end
```

No metatable of a metatable field, no table merging, no `__autocomplete` metamethod.

The only thing we need to do to make it work is to load `my_ee_module` after the `config` module. The order of loading is determined by the `lua_sources` array in `src/box/lua/init.c`. There is no problem to achieve it.